### PR TITLE
Refactor loader to decoupe from progressbar

### PR DESCRIPTION
**Pull request by chenkaiC4**
#87 

@NicolasRannou hi, I have refactored loader by decouping from progressbar.

Now loader.base extends event-emitter, and emits events at special hooks.

However, the current work need your review and suggestion, I don't want to merge this PR immediately, Let's base current work, and have more discussion. 😄 

Example at [fiddle](http://jsfiddle.net/chenkaiC4/v2u39mLh/)

### CODE REVIEW

**Comment by NicolasRannou** [comment]
Sounds good!

Couple of comments:
1. Do we need to include `events` in the package.json as it is part of `npm`? (https://nodejs.org/api/events.html#events_events)
Just setting the `npm` minium version should be enough right?

2. In the `event-progessbar` bar, would it make sense to also accept "ID" as "Dom Container"?
Renaming "DomContainer" to "DomTarget" and if it "DomTarget" is a string, get the related to element. If "DomTarget" is already a "Dom ellement", just pass it to the "DomContainer."

3. Rename `curTime` to `date` or `time`.

4. `parseing ` should be `parsing`

It looks good in general I'm really happy, will let you know when more comments after having deeper look!

**Comment by NicolasRannou** [comment]
5. I think we should make the event name consistent with the XMLHTTPREquest Standards (https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest)

`begin-fetch` -> `fetch-start`
`fetching` -> `fetch-progress`
`fetch-success` -> `fetch`

and also something like
`fetch-end` would be good, similar to https://developer.mozilla.org/en-US/docs/Web/Events/loadend


May not be too pretty but might be easier for people to use. WDYT?

**Comment by chenkaiC4** [comment]
@NicolasRannou thank you for advise.

yes, `event` is a inner package of `node.js`, i will remove it from the `package.json`. i'm glad to adhere to any standard 😄 , and give a more beautiful progress bar.

- [ ] remove `event-emitter` from `package.json`
- [ ] fix the event name according to xml standard
- [ ] adjust the selection logic of dom container 
- [ ] do a more beautiful progress bar (oh. css is my weakness 😢, though i always want to get it )

**Comment by NicolasRannou** [comment]
Thanks! Do not worry about the beautiful progress bar, I'm sure we can get something very cool.

Another thing that would be very useful (I know it is a lot to ask but we are now in the process of making AMI robust) would be to add testing to test all functions in the loader...

It implies:
1- Creating specs files such as: https://github.com/FNNDSC/ami/blob/dev/specs/models/models.base.spec.js
2- Test all functions (test for success and failure)

During development, I usually modify https://github.com/FNNDSC/ami/blob/dev/karma.conf.js#L16 and https://github.com/FNNDSC/ami/blob/dev/karma.conf.js#L27 to only run tests over the part I am working on.

Then `npm run test`.

WDYT?

Thanks!

**Comment by chenkaiC4** [comment]
@NicolasRannou Yes, Testing is necessary, I will give enough test case for loader. But what sample data i should use ? maybe files at `/data/dicom/` , i find settings at [karma.conf.js](https://github.com/FNNDSC/ami/blob/dev/karma.conf.js#L18)?

**Comment by NicolasRannou** [comment]
Yes exactly those files, that should be enough! Thanks!

**Comment by chenkaiC4** [comment]
@NicolasRannou ok! Let me have a try~ 😄

**Comment by chenkaiC4** [comment]
now the `fetch` events as follow(according to [XMLHttpRequest Draft](https://xhr.spec.whatwg.org/#events)):

`loadstart` => `fetch-start`

`progress` => `fetch-progress`

`error` => `fetch-error`

`load` => `fetch-success` [fetch success]

`timeout` => `fetch-timeout`

`loadend` => `fetch-end`  [The fetch completed (success or failure)]

`abort` => `fetch-abort`  [almost not appear here]

**Comment by NicolasRannou** [comment]
Cool! Sounds like you are getting there! :)

Sorry I keep adding things to the todo list but another thing I'd like is to change the name of the EventBasedProgressBar.

Ideally the filename and the class name would match, and I'd like to name it the other way around:
`src/helpers/helpers.event_progressbar.js` -> `src/helpers/helpers.progressbar.eventbased.js`

And the class name inside the `helpers.progressbar.eventbased.js` would be `HelpersProgressBarEventBased`.

The main reason for that is it allows all the `progressbar` helpers to be close alphabetically which is handy while development. Also it helps us know what the file does at first glance. Finally will be easy to group files together later on if that makes sense. (i.e. all progressbars in a `progressbar` directory)

WDYT?

**Comment by chenkaiC4** [comment]
@NicolasRannou just ok, i have written the tests, however need some helper, i will refresh the PR.

**Comment by NicolasRannou** [comment]
Looking really good @chenkaiC4 :)

**Comment by chenkaiC4** [comment]
@NicolasRannou thanks~

however, the event have not been test for now and test cases may not enough, suggestion may be well~  :)

----------------

phantomjs have been removed ?

```
16 03 2017 10:01:31.816:ERROR [plugin]: Error during loading "/home/travis/build/FNNDSC/ami/node_modules/karma-phantomjs-launcher" plugin:
  Path must be a string. Received null
16 03 2017 10:01:40.831:ERROR [karma]: Found 1 load error
```

**Comment by NicolasRannou** [comment]
Looking into the karma issue - I wanted to switch to `yarn` and didn't realize it would affect the pull request... on  it...

**Comment by chenkaiC4** [comment]
@NicolasRannou `yarn` is better than npm. Let's get it.

but, it seem the phantomjs not found.

**Comment by NicolasRannou** [comment]
Should be fixed now - re-building the pull request

File: src/helpers/helpers.js [hunk 1]
```diff
@@ -25,5 +25,5 @@ export default {
   VolumeRendering,
   Voxel,
   X,
-  EventBasedProgressBar,
+  ProgressBarEventBased,
```
Comment (NicolasRannou) [inline src/helpers/helpers.js]: @chenkaiC4  Can we move this one up close to ProgressBar? :)

Comment (chenkaiC4) [inline src/helpers/helpers.js]: ok~ :)

**Comment by NicolasRannou** [comment]
Finally, in the `specs/loaders/loaders.volume.spec.js`, we should remove the .tar suffix for all files. Ia used to be required but it is not anymore, that will make testing much better :)

After that I think we are very close to merging - need to review it all more in depth.

**Comment by chenkaiC4** [comment]
just remove the `.tar` ?  this will cause 'Not Found'~

**Comment by chenkaiC4** [comment]
if change the source file name, some test case at `pasrers.dicom.spec.js` should be modify.

**Comment by NicolasRannou** [comment]
Yes it is fine you can just remove the .tar in `loaders.volume.spec.js`, I already modified things in `parsers.dicom.spec.js`. When you will update the pull request, the Travis build (which run the tests will pass)

**Comment by NicolasRannou** [comment]
https://github.com/FNNDSC/ami/commit/640cde3bb31f5a3c6d79a92c0208f7f85e4d9d3a

**Comment by chenkaiC4** [comment]
@NicolasRannou 🥂 , more suggestions alway welcome.

i will take them later for now i have to go home :)

**Comment by chenkaiC4** [comment]
@NicolasRannou Hi, I have test the events sent by loader module using sinon spy. :)

**Comment by NicolasRannou** [comment]
Thanks @chenkaiC4 - I'll test it to make sure all demo and examples work then merge it!

**Comment by chenkaiC4** [comment]
@NicolasRannou cheer, if any something i need to do, just tell me, :)

File: specs/loaders/loaders.base.spec.js [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+          sinon.assert.calledWith(eventsHandleSpy['fetch-progress'], baseSinonMatch
+                                                                       .and(new sinon.match.hasOwn('total'))
+                                                                       .and(new sinon.match.hasOwn('loaded'))
+                                                                       );
+          // sinon.assert.calledWith(eventsHandleSpy['fetch-end'], baseSinonMatch);
+          done();
+        });
+    });
+
+    xit('the url is unavailbl', (done) => {
```
Comment (NicolasRannou) [inline specs/loaders/loaders.base.spec.js]: @chenkaiC4 typo :)

File: src/helpers/helpers.js [hunk 3]
```diff
@@ -24,4 +25,5 @@ export default {
   VolumeRendering,
   Voxel,
   X,
+  EventBasedProgressBar,
```
Comment (NicolasRannou) [inline src/helpers/helpers.js]: Can you please rename it ProgressBarEventBased and move it up alphabetically? @chenkaiC4

**Comment by NicolasRannou** [comment]
Added 2 minor comments @chenkaiC4 :)

**Comment by NicolasRannou** [comment]
Thanks @chenkaiC4 !

**Pull request merged**